### ISSUE TITLE
Failing test cases

### ISSUE THREAD

**Issue by jeffschriber** [evangelistalab/forte#13]
Here are the tests that fail:

        DSRG-MRPT2-6..............................................DOES NOT MATCH
        DSRG-MRPT3-5..............................................DOES NOT MATCH
        PCI-1.....................................................DOES NOT MATCH
        PCI-2.....................................................DOES NOT MATCH
        PCI-3.....................................................DOES NOT MATCH
        PCI-4.....................................................DOES NOT MATCH
        PCI-5.....................................................DOES NOT MATCH
        PCI-6.....................................................DOES NOT MATCH
        PCI-7.....................................................DOES NOT MATCH
        PCI-8.....................................................DOES NOT MATCH

Note that the DSRG-MRPT3-5 does not fail each time. When it does, it it is the 10th digit which causes the 9th digit (which is compared) to be different. 

For the DSRG-MRPT2-6, the number of trial vectors per root is 10, while it is 56 is the reference output. Maybe something changed with how this parameter is set.

**Comment by jeffschriber** [comment]
Update: I'm consistently getting four failed tests in the full test suite. They have been failing for a while: 

- [ ] DSRG-MRPT2-6
- [ ] DSRG-MRPT2-8-SA
- [ ] DSRG-MRPT3-3
- [ ] DSRG-MRPT3-4

**Comment by fevangelista** [comment]
I'll run them now.

**Comment by fevangelista** [comment]
Here is a summary of energy differences. These are small and @lcyyork attributes them to changes in psi4.

- DSRG-MRPT2-6
```
	CASSCF energy.....................................................PASSED
	DSRG-MRPT2 relaxed energy: computed value (-109.24618186850)
        does not match (-109.24612661379) to 10 digits.
```

- DSRG-MRPT3-3
```
	MCSCF energy......................................................PASSED
	DSRG-MRPT3 relaxed energy: computed value (-109.19516965)
        does not match (-109.19515717) to 7 digits.
```

- DSRG-MRPT3-4
```
	CASSCF energy.....................................................PASSED
	DSRG-MRPT3 relaxed energy: computed value (-108.97125543)
        does not match (-108.97125616) to 7 digits.
```

**Comment by lcyyork** [comment]
I think this is due to the changes of CASSCF. A direct evidence is that the frozen-core energy differs by 10^-4 between now and then. Other DSRG test cases either use MCSCF code (i.e. CASSCF(2,2)) or a more recent CASSCF code. I think it is OK to update the reference values.

**Comment by fevangelista** [comment]
Thanks @lcyyork. @jeffschriber: I don't have any problem with DSRG-MRPT2-8-SA. What error do you get? Can someone else check please.

**Comment by jeffschriber** [comment]
DSRG-MRPT2-8-SA is fine for me--I copied the wrong filename. dsrg-mrpt2-9-xms is failing for me instead.

**Comment by lcyyork** [comment]
Frozen-core energy is different. This is somewhat strange. DSRG-MRPT2-8 and -9 use the same SA-CASSCF(2,2) for LiF at 10.0 A, but they use difference frozen_docc, restricted_docc, and active for DSRG computations. I am not sure why the frozen-core energy for MRPT2-8 is almost perfect (10^-9) between versions of Jan 30 and Mar 7, but not for MRPT2-9 (10^-5).

**Comment by fevangelista** [comment]
I checked if it was connected to the different way we handle the core operator (we recently transitioned to using the DF_MP2 basis and not the DF_HF basis), but all these examples use conventional 4-e integrals.

**Comment by lcyyork** [comment]
For future convenience, I think it might be a good idea to add something like
C = wfn.Ca()
C.print_out()
to the input file or create another file called Ca.ref to compare CASSCF optimized orbitals from version of version.

**Comment by fevangelista** [comment]
Good point.

**Comment by fevangelista** [comment]
@lcyyork: From a conversation with @dgasmith it looks like it might be an issue related to canonicalization of the MOs.
The components required to build the average Fock matrix are available from the `mcscf_obj` in `mcscf_solver.py` using the following functions:
```
    SharedMatrix current_AFock() { return matrices_["AFock"]; }
    SharedMatrix current_IFock() { return matrices_["IFock"]; }
```
We should be able to easily add an option to canonicalize the MOs. When we make a pull request we should perhaps combine this with our code to orthonormalize the basis when orbitals are read from a previous computation.

**Comment by lcyyork** [comment]
That is fine. I have the orthonormalization/semicanonicalization code on the cluster written in python.

**Comment by lcyyork** [comment]
So I run dsrg-mrpt2-6 again using the previous version of PSI4 (b2d0904b4... be specific), which gives me a pass. All energies match exactly with the reference values. I will add the corresponding Ca shortly.

**Comment by fevangelista** [comment]
Thanks York. How do I toggle the semicanonicalization of the orbitals in your dsrg-mrpt2 code? I have added code to psi4 and would like to test it with Forte.

**Comment by lcyyork** [comment]
The dsrg-mrpt2 code itself will tests the Fock matrix (exclude frozen core orbitals). If its diagonal blocks are diagonal, it will simply run. If they are not diagonal, the dsrg-mrpt2 code will automatically generate the rotation matrix, which is used to transform the integrals and amplitudes, and back transform the amplitudes. So I guess there is no way to control the semicanonicalization from forte options. There is a object called something like ignore_semi, which will make the code skip generating the rotation matrix.

As for testing the CASSCF code, I think the input of dsrg-mrpt2-6 is enough. Note that after psi4 casscf run, I use the casci code of forte to semicanonicalize the orbitals. The trick is to use exactly the same frozen_docc, restricted_docc, active, etc. in this block as the psi4 casscf block. For this test case, there is no frozen orbitals in casscf. After semicanonicalization, the real dsrg computation will freeze the two core orbitals.

So I think you can test the casscf code by disabling the middle block of dsrg-mrpt2-6, and the same orbitals (given by C_semi.ref) should be obtained.

**Comment by fevangelista** [comment]
Thanks @lcyyork. I was able to test the new code in Psi4 by disabling the call to CASSCF in the middle of the test. Once the code in Psi4 is merged we can update all tests that freeze core.

**Comment by lcyyork** [comment]
That is great @fevangelista. I am not sure you should include frozen orbitals to the mcscf_solver or not.

**Comment by lcyyork** [comment]
Or maybe frozen orbitals are already been taken care of. My bad.

**Comment by fevangelista** [comment]
What I do now is semicanonicalize all the restricted_docc CASSCF MOs. frozen_docc MOs in CASSCF are not semicanonicalized because I currently don't have access to that block of the Fock matrix. I think other codes do things this way (Molpro?).

**Comment by fevangelista** [comment]
@lcyyork: Now that Psi4 can semicanonicalize CASSCF MOs, can you please remove the semicanonicalization step in the test cases that need it?

**Comment by fevangelista** [comment]
Closed with commit ea7d136862974a66f21c1e208c6ee9bff0556001.
